ci-operator/step-registry/cucushift/upgrade/check/capability/cucushift-upgrade-check-capability-commands.sh

```diff
@@ -0,0 +1,451 @@
+declare -A caps_operator
+caps_operator[baremetal]="baremetal"
+caps_operator[marketplace]="marketplace"
+caps_operator[openshift-samples]="openshift-samples"
+caps_operator[CSISnapshot]="csi-snapshot-controller"
+caps_operator[Console]="console"
+caps_operator[Insights]="insights"
+caps_operator[Storage]="storage"
+caps_operator[NodeTuning]="node-tuning"
+caps_operator[MachineAPI]="machine-api control-plane-machine-set cluster-autoscaler"
+caps_operator[ImageRegistry]="image-registry"
+caps_operator[OperatorLifecycleManager]="operator-lifecycle-manager operator-lifecycle-manager-catalog operator-lifecycle-manager-packageserver"
+
+# Mapping between optional capability and resources
+# Need update when new resource marks as optional
+caps_resource_list="Build DeploymentConfig"
+declare -A caps_resource
+caps_resource[Build]="builds buildconfigs"
+caps_resource[DeploymentConfig]="deploymentconfigs"
+
+# Initialize the variables
+declare -A migrated_caps=()
+if ! extract_migrated_caps "${version_set[-1]}" "${version_set[0]}"; then
+    echo "ERROR: extract_migrated_caps failed" && exit 1
+fi
+
+expected_enabled_caps=""
+expected_implicit_caps=""
+expected_disabled_caps=""
+
+source_baseline_caps=$(get_caps_for_version_capset "${version_set[-1]}" "${baselinecaps_from_cluster}")
+target_baseline_caps=$(get_caps_for_version_capset "${version_set[0]}" "${baselinecaps_from_cluster}")
+
+source_known_caps=$(get_caps_for_version_capset "${version_set[-1]}" "vCurrent")
+target_known_caps=$(get_caps_for_version_capset "${version_set[0]}" "vCurrent")
+
+new_caps=$(get_new_caps "${source_known_caps}" "${target_known_caps}")
+
+additionalcaps_from_cluster=$(oc get clusterversion version -ojson | jq -r '.spec.capabilities.additionalEnabledCapabilities[]')
+
+echo -e "\nsource_baseline_caps is ${source_baseline_caps}\
+\ntarget_baseline_caps is ${target_baseline_caps}\
+\nsource_known_caps is ${source_known_caps}\ntarget_known_caps is ${target_known_caps}\
+\nnew_caps is ${new_caps}\nadditionalcaps_from_cluster is ${additionalcaps_from_cluster}"
+
+# vCurrent baseline set should grow new caps of target version explicitly after upgrade
+if [[ "${baselinecaps_from_cluster}" == "vCurrent" ]]; then
+    expected_enabled_caps="${additionalcaps_from_cluster} ${target_baseline_caps}"
+fi
+
+# v4.x baseline set should grow new caps in v4.x of target version explicitly after upgrade
+# v4.x baseline set should grow new caps in target version implicitly
+# v4.x baseline set should grow new caps implicitly after upgrade if cap migrates to a cap out of v4.x & additional caps of target version
+if [[ "${baselinecaps_from_cluster}" == "v4."* ]]; then
+    source_enabled_caps="${source_baseline_caps} ${additionalcaps_from_cluster}"
+    target_enabled_caps="${target_baseline_caps} ${additionalcaps_from_cluster}"
+    expected_enabled_caps="${target_enabled_caps}"
+    
+    expected_enabled_caps="${expected_enabled_caps} ${new_caps}"
```

Comment:
From my understanding, the expected enabled caps is the source enabled caps in  + new_caps, sounds the current logic would make several duplicated caps show up in the the expected enabled caps, right?

Reply:
Here `d` is the key point, generally I think v4.x would have no any change on both source and target version, of course, machine-api is an exception change. 

So let us take 4.13 + v4.11as an example source cluster, now it is going to be upgraded to 4.14.

In source cluster, baremetal, marketplace, openshift-sample is the enabled optional caps, the new cap is MachineAPI, Build, DeploymentConfig, ImageRegistry. 

So the target enabled should be baremetal, marketplace, openshift-sample + MachineAPI, Build, DeploymentConfig, ImageRegistry.

In your example, the key point is the same v4.x has different sets in source and target, and the different cap - `d` is also never show up in the new cap, but till now, I do not figure out such example yet.

Reply:
Yeah, my example doesn't happen yet. I'm just in case it occurs some day and having `d` there twice won't hurt because I remove duplicates later